actionview/lib/action_view/helpers/tag_helper.rb

```diff
@@ -344,6 +344,8 @@ def tag(name = nil, options = nil, open = false, escape = true)
       #   <% end -%>
       #    # => <div class="strong">Hello world!</div>
       def content_tag(name, content_or_options_with_block = nil, options = nil, escape = true, &block)
+        raise ArgumentError, "Invalid HTML tag name: #{name}" unless /\A[a-zA-Z][\w-]*\z/.match?(name)
```

Comment:
Why is this check here, instead of in `TagBuilder#content_tag_string`? Example test I would expect to pass:

```ruby
  assert_raise(ArgumentError) { tag.send("my@element") }
```

Or maybe put another way: why is `content_tag` the only method that's doing this check, instead of all the tag helpers? Just trying to understand the problem we're trying to solve here.

Reply:
Also note that there are other unprotected code paths:

- `TagBuilder#content_tag_string` mentioned above
- `tag` has a code path that doesn't use `TagBuilder#content_tag_string`
- `TagBuilder#tag_string` has a code path that doesn't use `TagBuilder#content_tag_string` (this method is used for `method_missing`)

Reply:
I would like to see the check wrapped up as a (private?) method, too, could be as simple as

```
  def ensure_valid_html5_tag_name(name)
    raise ArgumentError, "Invalid HTML5 tag name: #{name}" unless /\A[a-zA-Z][\w-]*\z/.match?(name)
  end
```

Reply:
We decided to introduce a validation in content_tag due to a discussion on https://github.com/rails/rails/pull/49120#discussion_r1315644911. The discussion revolved around resolving the issue where using `wrapper_tag: ""` (with an empty string) option in `simple_format` resulted in an empty tag `<></>`. As a result of this discussion, it was determined that having validation in `content_tag` would help address such issues.

Reply:
I'm updating code based on your suggession

Reply:
@flavorjones Right now we are escaping some characters, do we need the validation before escaping?

Reply:
> /\A[a-zA-Z][^\s/>]+\z/

This regex causing a premature end of char-class error.

Reply:
> I not positive, but I don't think we should escape the name if we're doing the validation. For example:
> 
> ```ruby
> ERB::Util.xml_name_escape("asdf-♥")
> => "asdf-_"
> ```

You meant to say, we can completely remove the escape functionality from all the tags?

Reply:
> > /\A[a-zA-Z][^\s/>]+\z/
>
>This regex causing a premature end of char-class error.

Right, so please escape the solidus character: `/\A[a-zA-Z][^\s\/>]+\z/`

```ruby
re = /\A[a-zA-Z][^\s\/>]+\z/

re.match? "asdf"
# => true
re.match? "asdf-asdf"
# => true
re.match? "asdf-♥"
# => true
re.match? "asdf asdf"
# => false
re.match? "asdf/asdf"
# => false
```

> You meant to say, we can completely remove the escape functionality from all the tags?

Yes, if we are raising an exception on invalid tag names, then we no longer need to bother escaping invalid characters in that invalid tag name. Make sense?

I'm happy to jump on a huddle/hangout/zoom/phone call to chat in realtime if that would be helpful! I appreciate the work you've put into this.

Reply:
Thanks @flavorjones. Lemme know you available time for a call :)

Reply:
OK, change the `+` to a `*` please ... my original comment was

> I think the regex should be something like: /\A[a-zA-Z][^\s/>]+\z/

My intention was to point you in the general direction of a solution, not to be explicit, so I apologize for the confusion. It seems like you know what's going on now?

Reply:
No apologize please. Now it makes sense to me. Thanks for your great feedback on this.